persist/fs/commitlog/reader.go

```diff
@@ -109,58 +192,267 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		err := r.startBackgroundWorkers()
+		if err != nil {
+			return Series{}, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), err
+		}
+	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream. I.E with a
+	// concurrency of two then we'll start with an initial structure like this:
+	// 		r.outBufs:
+	// 			[0]: []
+	// 			[1]: []
+	// After several reads from the commitlog we end up with:
+	// 		r.outBufs:
+	// 			[0]: [a, c]
+	// 			[1]: [b, d]
+	// If we now read from the outBufs in round-robin order (0, 1, 0, 1) we will
+	// end up with the correct global ordering: a, b, c, d
+	rr, ok := <-r.outBufs[r.nextIndex%r.numConc]
+	if !ok {
+		return Series{}, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), io.EOF
+	}
+	if rr.pendingMetadata != nil {
+		// Wait for metadata to be available
+		rr.series, rr.resultErr = rr.pendingMetadata.wait()
 	}
+	r.nextIndex++
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok && len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) startBackgroundWorkers() error {
+	// Make sure background workers are never setup more than once
+	set := atomic.CompareAndSwapInt64(&r.bgWorkersInitialized, 0, 1)
+	if !set {
+		return errCommitLogReaderMultipleReadloops
+	}
+
+	// Start background worker goroutines
+	go r.readLoop()
+	for i, decoderBuf := range r.decoderBufs {
+		localDecoderBuf, outBuf := decoderBuf, r.outBufs[i]
+		go r.decoderLoop(localDecoderBuf, outBuf)
+	}
+
+	return nil
+}
+
+func (r *reader) readLoop() {
+	index := 0
+	defer r.shutdown()
+	for {
+		select {
+		case <-r.cancelCtx.Done():
 			return
+		default:
+			data, err := r.readChunk()
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
+			if err == io.EOF {
+				return
+			}
+		}
+	}
+}
+
+func (r *reader) shutdown() {
+	for _, decoderBuf := range r.decoderBufs {
+		close(decoderBuf)
+	}
+	r.shutdownCh <- r.close()
+}
+
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
```

Comment:
This function is doing a lot, do you think you can break it into smaller parts to make it easier to follow ~ for e.g. the metadata retrieval bits on line 344-364 can probably be another function on the `reader`, same for the pending metadata release from lines 320-344.